[WIP] DOCS-1043: Added topics relating to ALP (#257)
tigera/docs: calico-enterprise/network-policy/application-layer-policies/alp.mdx

---
description: Enforce application layer policies in your cluster to limit access controls based on L7 attributes.
---

# Enable and enforce application layer policies

Application layer policies let you limit access controls based on L7 attributes. 

## Before you begin

### Unsupported
{{prodname}} implements application layer policy using Envoy as a DaemonSet. This means you cannot use application layer policy alongside a service mesh like Istio.

### Limitations
- Application layer policy supports restricting only ingress traffic

> - Support for L7 attributes are limited to HTTP method, and URL exact/prefix path 

Reviewer: omit comma?

HTTP method, and URL exact/prefix path

Author: Done

- Supported protocols are limited to TCP-based protocols (for example, HTTP, HTTPS, or gRPC)
- You can control application layer policies only at the cluster level (not per namespace) 

## How to
- [Enable application layer policies](#enable-application-layer-policies)
- [Enforce application layer policies for ingress traffic](#enforce-application-layer-policies-for-ingress-traffic)
- [Disable application layer policies](#disable-application-layer-policies)

### Enable application layer policies (ALP)
In the ApplicationLayer custom resource, set the `applicationLayerPolicy` field to Enabled.

> ```                                                                                           

Reviewer: ```yaml

Author: @bmckercher123 I'm unable to see the change here.

apiVersion: operator.tigera.io/v1
kind: ApplicationLayer
metadata:
  name: tigera-secure 
spec:
  applicationLayerPolicy: Enabled
  
```

### Enforce application layer policies for ingress traffic

You can restrict ingress traffic using HTTP match criteria using Network policy or Global network policy. 
For a list of all HTTP match parameters, see [Network policy](/reference/resources/networkpolicy.mdx) and [Global network policy](/reference/resources/globalnetworkpolicy.mdx).

In the following example, the trading app is allowed ingress traffic only for HTTP GET requests that match the exact path /projects/calico, or that begins with the prefix, /users.

```
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: customer
spec:
  selector: app == 'tradingapp'
  ingress:
   - action: Allow
     http:
       methods: ["GET"]
       paths:
         - exact: "/projects/calico"
         - prefix: "/users"
  egress:
    - action: Allow
```

> ### Disable application layer policies

Reviewer: Align...title has "policies" and first sentence has "policy".

Author: Done

---
description: Enforce application layer policies in your cluster to limit access controls based on L7 attributes.
---

# Enable and enforce application layer policies

Application layer policies let you limit access controls based on L7 attributes. 

## Before you begin

### Unsupported
{{prodname}} implements application layer policy using Envoy as a DaemonSet. This means you cannot use application layer policy alongside a service mesh like Istio.

### Limitations
- Application layer policy supports restricting only ingress traffic

> - Support for L7 attributes are limited to HTTP method and URL path 

Reviewer 2: ```suggestion
- Support for L7 attributes are limited to HTTP method, and URL exact/prefix path 
```

> - Support for protocols is limited to TCP-based protocols (for example, HTTP, HTTPS, or gRPC)

Reviewer 2: ```suggestion
- Supported protocols are limited to TCP-based protocols (for example, HTTP, HTTPS, or gRPC)
```

- You can control application layer policies only at the cluster level (not per namespace) 

## How to
- [Enable application layer policies](#enable-application-layer-policies)
- [Enforce application layer policies for ingress traffic](#enforce-application-layer-policies-for-ingress-traffic)

## Enable application layer policies (ALP)
In the ApplicationLayer custom resource, set the `applicationLayerPolicy` field to Enabled.

```
kubectl apply -f - <<EOF                                                                                           
apiVersion: operator.tigera.io/v1
kind: ApplicationLayer
metadata:
  name: tigera-secure 
spec:
  applicationLayerPolicy: Enabled
EOF
```

> Use the same field to disable the application layer policy.

Reviewer 2: ```suggestion

Performing the following will disable Applicaition Layer Policy:

1. Setting the `applicationLayerPolicy` field in the `ApplicationLayer` custom resource to `Disabled`, or
1. Removing the `applicationLayerPolicy` field entirely, or
1. Removing the whole ApplicationLayer` custom resource

See note further below for more details. 
```

Author: I rephrased a bit :).

To disable the policy, do one of the following steps:
 - Set the `applicationLayerPolicy` field in the `ApplicationLayer` custom resource to `Disabled`.
 - Remove the `applicationLayerPolicy` field entirely.
 - Delete the ApplicationLayer` custom resource.
